src/code42cli/cmds/auditlogs.py

```diff
@@ -0,0 +1,255 @@
+        affected_usernames=affected_username,
+    )
+
+
+@audit_logs.command()
+@filter_options
+@checkpoint_option
+@server_options
+@send_to_format_options
+@sdk_options()
+def send_to(
+    state,
+    hostname,
+    protocol,
+    format,
+    begin,
+    end,
+    event_type,
+    username,
+    user_id,
+    user_ip,
+    affected_user_id,
+    affected_username,
+    use_checkpoint,
+):
+    """Send audit logs to the given server address."""
+    save_checkpoint = None
+    if use_checkpoint:
+        cursor = _get_audit_log_cursor_store(state.profile.name)
+        save_checkpoint = lambda ts: cursor.replace(use_checkpoint, ts)
+        if not begin:
+            begin = cursor.get(use_checkpoint)
+    
+    _send_to(
+        state.sdk,
+        hostname,
+        protocol,
+        format,
+        save_checkpoint,
+        begin_time=begin,
+        end_time=end,
+        event_types=event_type,
+        usernames=username,
+        user_ids=user_id,
+        user_ip_addresses=user_ip,
+        affected_user_ids=affected_user_id,
+        affected_usernames=affected_username,
+    )
+
+
+def _search(sdk, format, save_checkpoint, **filter_args):
+    formatter = OutputFormatter(format, AUDIT_LOGS_DEFAULT_HEADER)
+    response_gen = sdk.auditlogs.get_all(**filter_args)
+    events = _get_all_audit_log_events(response_gen)
+    event_count = len(events)
+    if not event_count:
+        click.echo("No results found.")
+        return
+    elif event_count > 10:
+        click.echo_via_pager(formatter.get_formatted_output(events))
+    else:
+        formatter.echo_formatted_list(events)
+    if save_checkpoint:
+        ts = _parse_audit_log_timestamp_string_to_timestamp(events[0]["timestamp"])
+        save_checkpoint(ts)
+
+
+def _send_to(sdk, hostname, protocol, format, save_checkpoint, **filter_args):
+    logger = get_logger_for_server(hostname, protocol, format)
+    with warn_interrupt():
+        response_gen = sdk.auditlogs.get_all(**filter_args)
+        events = _get_all_audit_log_events(response_gen)
+        if not events:
+            click.echo("No results found.")
+            return
+        for event in reversed(events):
+            logger.info(event)
+            if save_checkpoint:
+                ts = _parse_audit_log_timestamp_string_to_timestamp(event["timestamp"])
+                save_checkpoint(ts)
+
+
+def _get_all_audit_log_events(response_gen):
+    events = []
+    try:
+        for response in response_gen:
+            response_dict = json.loads(response.text)
+            if EVENT_KEY in response_dict:
+                events.extend(response_dict.get(EVENT_KEY))
+    except KeyError:
+        # API endpoint (get_page) returns a response without events key when no records are found
+        # e.g {"paginationRangeStartIndex": 10000, "paginationRangeEndIndex": 10000, "totalResultCount": 1593}
+        pass
+    return events
+    
+
+def _parse_audit_log_timestamp_string_to_timestamp(ts):
```

Comment:
Won't this potentially cause us to miss events if we bump ms but already more events exist on the backend so now we list the end of the last ms of events?

Reply:
the way the extractor handles this is to exclude the highest timestamp values from being retrieved and add 1ms to the _second_ highest timestamp.

Reply:
We are going to have to port that logic over eventually anyway, maybe now is the time...

Reply:
So does any given extractor run always leave at least one event out? 

Since we're going to be pulling the checkpointing logic fully into the CLI when we remove the extractor, I wonder if we should update the cursor to also store a(n) event identifier(s) alongside the timestamp, that way we can always process all the events at the end of each run, but also dedupe events at the start of each run.

Reply:
Wondering why don't we use the API filter option to implement checkpoint behaviour. The audit-log [API](https://support.code42.com/Administrator/Cloud/Monitoring_and_managing/Code42_API_resources/Search_Audit_Log_events_with_the_Code42_API) supports filter by date-range. We can pass the checkpoint value in the start-time field in the api request.

Reply:
That is what we're doing, we replace the `begin` argument with the checkpoint value if `--use-checkpoint` is given and a timestamp exists in the cursor: 

https://github.com/code42/code42cli/pull/169/files#diff-ac4db69e66795de05d5b7b5d999d10611f3cefe5f220cb6ba1442b9abe751714R132-R138

Reply:
I've implemented audit logs checkpointing logic to store hashes of the last events of any given run that share the same timestamp, then the next run's events are hashed and compared and pruned if they exist in the checkpoint.